File: email-feeder-go/internal/pipeline/caps.go
package pipeline

import "errors"

var ErrCapsExceeded = errors.New("submission exceeds caps")

type Caps struct {
	MaxAttachmentBytes int
	MaxAttachments     int
	MaxTotalBytes      int
}

func (c Caps) check(p *Parsed) error {
	total := len(p.Wrapper)
	if len(p.Attachments) > c.MaxAttachments {
		return ErrCapsExceeded
	}
	for _, a := range p.Attachments {
		if len(a.Content) > c.MaxAttachmentBytes {
			return ErrCapsExceeded
		}
		total += len(a.Content)
	}
	if total > c.MaxTotalBytes {
		return ErrCapsExceeded
	}
	return nil
}

File: email-feeder/classes/services/caps.py
"""Submission size/count caps — bound memory/disk against hostile mail."""


class CapsExceeded(Exception):
    """Raised when a submission exceeds the configured caps."""


def check_caps(attachments, wrapper_bytes: int, caps) -> None:
    if len(attachments) > caps.max_attachments:
        raise CapsExceeded(
            f"{len(attachments)} attachments > max {caps.max_attachments}")
    total = wrapper_bytes
    for att in attachments:
        size = len(att.content or b"")
        if size > caps.max_attachment_bytes:
            raise CapsExceeded(
                f"attachment {size} bytes > max {caps.max_attachment_bytes}")
        total += size
    if total > caps.max_total_bytes:
        raise CapsExceeded(f"total {total} bytes > max {caps.max_total_bytes}")
